### PR TITLE
Express LTS strategy

### PR DESCRIPTION
**Pull request by kibertoad**
refs #196

### CODE REVIEW

File: docs/LTS-strategy.md
Comment (wesleytodd) [inline docs/LTS-strategy.md:8]: I think folks might read this as "in order of priority". I know that is not intended, but I wonder if we should move "Provide good experience for users who want to keep their Express projects that up-to-date" to the top?

File: docs/LTS-strategy.md
Comment (wesleytodd) [inline docs/LTS-strategy.md:13]: Love this section, especially this line! 🎉

File: docs/LTS-strategy.md
Comment (wesleytodd) [inline docs/LTS-strategy.md:20]:
```suggestion
* Keep options open for adopting newer patterns and tooling
```

File: docs/LTS-strategy.md
Comment (wesleytodd) [inline docs/LTS-strategy.md:27]: I think this is the section I want us to discuss more deeply. Ideally we also put together a calendar view like Node does to see in comparison to Node majors how express majors would line up.

File: docs/LTS-strategy.md
Comment (wesleytodd) [inline docs/LTS-strategy.md:35]:
```suggestion
* Several semver major branches will be maintained at the same time while we have overlapping supported versions
```

File: docs/LTS-strategy.md
Comment (wesleytodd) [inline docs/LTS-strategy.md:36]: Thumbs up on this, but I do think maybe we want to discuss this as a group in more detail.

File: docs/LTS-strategy.md
Comment (wesleytodd) [inline docs/LTS-strategy.md:37]:
```suggestion
* Critical bugfixes will be backported to older supported semver major branches
```

I think we should commit to this.

File: docs/LTS-strategy.md
Comment (wesleytodd) [inline docs/LTS-strategy.md:38]:
```suggestion
* Security fixes will be backported to older supported semver major branches
```

I also think we should commit to doing this for their support lifetime.

**Review by wesleytodd — CHANGES_REQUESTED** [review]
This is AWESOME. Thanks for getting this started. I made some comments to seed conversation, and I think we might want to expand a bit more on some of them, but generally I think this is an awesome start.

File: docs/LTS-strategy.md
Comment (ljharb) [reply docs/LTS-strategy.md:13]: I'm not actually sure this makes sense - a healthy balance imo is the same thing as "widest possible set". If a node version prevents using modern toolchains, for example, then that's not a version that'd be included in that set.

File: docs/LTS-strategy.md
Comment (ljharb) [inline docs/LTS-strategy.md:29]: this is great, thanks for capturing it

File: docs/LTS-strategy.md
Comment (ljharb) [inline docs/LTS-strategy.md:30]:
```suggestion
* Version support removal must come with clear benefits for one of the three - Express users, Express maintainers, or the Node.js core project. Version support is not dropped on a merely time-based basis, or otherwise arbitrarily.
```

File: docs/LTS-strategy.md
Comment (ljharb) [reply docs/LTS-strategy.md:35]: this should probably be done as a matter of course anyways, so that if a security issue pops up that necessitates a backport, all the necessary infra is already in place.

File: docs/LTS-strategy.md
Comment (ljharb) [reply docs/LTS-strategy.md:37]: i agree, but i think there should also be "may be backported to unsupported semver major branches" as well.

**Review by ljharb — COMMENTED** [review]
Thanks for writing this up!

File: docs/LTS-strategy.md
Comment (kibertoad) [reply docs/LTS-strategy.md:13]: @ljharb I'll think about possible wording improvements here, but the intent is to emphasize that we only want to support a set of versions that do not contradict the rest of the goals defined. This is very important to make explicit in the context of Express specifically, as it has a long-standing tradition of preserving the compatibility to the detriment of everything else.

File: docs/LTS-strategy.md
Comment (UlisesGascon) [reply docs/LTS-strategy.md:27]: Yeah! Having a calendar is one of the best ways to explain the rules for the end users, but currently is very solid 🙂

File: docs/LTS-strategy.md
Comment (UlisesGascon) [reply docs/LTS-strategy.md:13]: +1 this is a great way to explain it

File: docs/LTS-strategy.md
Comment (UlisesGascon) [reply docs/LTS-strategy.md:36]: Yep, we should backport only fixes, perf improvements and so on.. but agree that new features are out of scope 💯

**Review by UlisesGascon — APPROVED** [review]
Fantastic proposal @kibertoad! Thanks for doing this. 

IMO this is great and I feel confident with the current state (previous discussed comments aside). I assume that might do more tweaks in the future (but not a blocker for this PR), specially when we will include the release schedule (calendar) as we will need to fit this ideas more intro a concrete schematics 👍

File: docs/LTS-strategy.md
Comment (ljharb) [reply docs/LTS-strategy.md:13]: I don’t think any of the detriments you refer to have anything to do with refusing to drop compat for node, but with other things - but i look forward to an alternative phrasing.

Comment (wesleytodd) [reply docs/LTS-strategy.md:13]: > a healthy balance imo is the same thing as "widest possible set"

As other said, I disagree with the wording but maybe your meaning hinges on the definition we use for "possible"?

Like if we said "we want to ship diagnostic channels so it is not possible to support versions before them" (I am trying to make a hypnotically based on a reality from the past https://github.com/pillarjs/router/pull/96) does that fit your definition? We would drop the old node versions which dont support it in the next major so we could ship support without a lot of compatibility code.

File: docs/LTS-strategy.md
Comment (wesleytodd) [reply docs/LTS-strategy.md:27]: I think we should make a diagram like node has, maybe even overlay our support *over* the existing node support schedules?

File: docs/LTS-strategy.md
Comment (ljharb) [reply docs/LTS-strategy.md:13]: Yes, of course it fits my definition. If we need a feature that requires dropping a version, then that version is no longer possible to support.

I also especially like the implied path of "ship support with compat code, and then drop the compat code and the node versions that need it in the next major" - that's a *very* user-friendly way to do a major bump.

**Comment by wesleytodd** [comment]
Here is an example we can also reference. https://github.com/debug-js/debug/releases?page=2

Debug, a direct dep, dropped node versions which blocked us updating.

File: docs/LTS-strategy.md
Comment (kibertoad) [reply docs/LTS-strategy.md:7]: addressed

File: docs/LTS-strategy.md
Comment (kibertoad) [reply docs/LTS-strategy.md:27]: @wesleytodd @UlisesGascon I think @ljharb strongly objected to having a timed deprecation without a practical justification. We need to have this established unambiguously within the strategy. Do we or not do time schedule-based Node version deprecation?

File: docs/LTS-strategy.md
Comment (kibertoad) [reply docs/LTS-strategy.md:36]: let's revise in a separate PR if needed

File: docs/LTS-strategy.md
Comment (kibertoad) [inline docs/LTS-strategy.md:14]: @ljharb @wesleytodd @UlisesGascon Please review the updated wording

Comment (ljharb) [reply docs/LTS-strategy.md:14]: How about this:
```suggestion
* Prioritizing "support older Node.js versions" over other factors. We aim to maintain a healthy balance between backwards compatibility and toolchain modernization
```

Comment (kibertoad) [reply docs/LTS-strategy.md]: sure, that works!

**Review by ljharb — COMMENTED** [review]
LGTM! thanks for iterating

**Comment by kibertoad** [comment]
OK, I think this might be ready to be merged as a first version, but of course we should keep iterating over it.

Should we open the discussion of what toolchain to use as a next step? This would help inform the decision of which particular versions to support in the first upcoming semver major release

File: docs/LTS-strategy.md
Comment (wesleytodd) [reply docs/LTS-strategy.md:27]: Can we merge without a consensus on this? I am leaning yes? And then we can open targeted PRs to update these and discuss after?

**Review by wesleytodd — APPROVED** [review]

File: docs/LTS-strategy.md
Comment (ljharb) [reply docs/LTS-strategy.md:27]: If it’s worded as something like “if a release is warranted, this is the schedule when it will happen - but if not, no release will happen” then my concern is gone.

Comment (ljharb) [reply docs/LTS-strategy.md:27]: Either way, but it seems easy enough to do it before merging ¯\\\_(ツ)_/¯

Comment (wesleytodd) [reply docs/LTS-strategy.md:27]: @kibertoad if you get to it, awesome. If not, lets try and merge this before EOD so we can start those other PRs to clarify and expand.

Comment (kibertoad) [reply docs/LTS-strategy.md:27]: addressed. unless there are major outstanding objections to some of the wording, I would suggest merging it as-is and iterate in follow-up PRs.

@wesleytodd Can I get push permissions for the repo?

Comment (kibertoad) [reply docs/LTS-strategy.md:27]: thanks!

**Pull request merged**
